selfdrive/pandad/pandad.cc

```diff
@@ -546,9 +516,38 @@ void peripheral_control_thread(Panda *panda, bool no_fan_control) {
   }
 }
 
-void pandad_main_thread(std::vector<std::string> serials) {
-  LOGW("launching pandad");
+void pandad_run(std::vector<Panda *> &pandas) {
+  const bool no_fan_control = getenv("NO_FAN_CONTROL") != nullptr;
+  const bool spoofing_started = getenv("STARTED") != nullptr;
+  const bool fack_send = getenv("FAKESEND") != nullptr;
+
+  PubMaster pm({"can", "pandaStates", "peripheralState"});
+  RateKeeper rk("pandad", 100);  // 100 hz
+
+  while (!do_exit && check_all_connected(pandas)) {
+    can_recv(pandas, &pm);
+    can_send(pandas, fack_send);
```

Comment:
can_send will need to remain in another thread for now, due to the following case:

while the panda TX buffer is full, it will NACK the SPI messages that try to add more messages: https://github.com/commaai/openpilot/blob/8f3fc699d4e06811fd29fbf64fe89032cce563b9/selfdrive/pandad/spi.cc#L238.

Reply:
FYI, you won't notice this case in normal use, but there are edge cases when it can happen (i.e. the cable is faulty and doesn't have a good connection to the CAN bus)